src/main/java/org/spongepowered/common/world/level/chunk/storage/SpongeChunkEntities.java

```diff
@@ -0,0 +1,233 @@
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+package org.spongepowered.common.world.level.chunk.storage;
+
+import com.google.common.collect.ImmutableList;
+import net.minecraft.util.Tuple;
+import net.minecraft.world.level.Level;
+import org.checkerframework.checker.nullness.qual.MonotonicNonNull;
+import org.checkerframework.checker.nullness.qual.Nullable;
+import org.spongepowered.api.data.persistence.DataContainer;
+import org.spongepowered.api.entity.Entity;
+import org.spongepowered.api.entity.EntityType;
+import org.spongepowered.api.entity.living.player.Player;
+import org.spongepowered.api.util.AABB;
+import org.spongepowered.api.world.chunk.EntityChunk;
+import org.spongepowered.api.world.volume.stream.StreamOptions;
+import org.spongepowered.api.world.volume.stream.VolumeStream;
+import org.spongepowered.common.bridge.world.level.LevelBridge;
+import org.spongepowered.common.util.VecHelper;
+import org.spongepowered.common.world.storage.SpongeChunkLayout;
+import org.spongepowered.common.world.volume.VolumeStreamUtils;
+import org.spongepowered.common.world.volume.buffer.entity.ObjectArrayMutableEntityBuffer;
+import org.spongepowered.math.vector.Vector3d;
+import org.spongepowered.math.vector.Vector3i;
+
+import java.util.AbstractMap;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.List;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.UUID;
+import java.util.function.Predicate;
+import java.util.stream.Stream;
+
+public final class SpongeChunkEntities implements EntityChunk {
+
+    private final Level level;
+    private final Vector3i chunkPosition;
+    private final Stream<net.minecraft.world.entity.Entity> entities;
+
+    private @MonotonicNonNull SpongeChunkLayout chunkLayout;
+    private @MonotonicNonNull Vector3i blockMin;
+    private @MonotonicNonNull Vector3i blockMax;
+    private @MonotonicNonNull List<net.minecraft.world.entity.Entity> newEntities;
+
+    public SpongeChunkEntities(final Level chunkLayout, final Vector3i chunkPosition, final Stream<net.minecraft.world.entity.Entity> entities) {
```

Comment:
`chunkLayout` -> `level`

Reply:
Also should be `ServerLevel` as this would never be client-side (unless you are seeing something here I'm not seeing).